### ISSUE TITLE
Protected reduce_sum function names

### ISSUE THREAD

**Issue by aornugent** [stan-dev/stanc3#906]
#### Summary:
I discovered the hard way that TBB doesn't like functions that end in `_lp`. 

#### Description:
I was searching for something to call a function to paralellise and landed on `partial_lp` (ignoring completely the correctness of this name). The error below left me scratching my head when compilation failed.

Luckily the model had worked prior to the name change so I was able to isolate the issue. Renaming the function `partial_ll` compiles and runs fine. Could this be checked by the parser to return a better error message?

#### Reproducible Steps:
Compile this tweaked reduce_sum example:

```stan
functions {
  real partial_sum_lp(int[] slice_n_redcards,
                        int start, int end,
                        int[] n_games,
                        vector rating,
                        vector beta) {
    return binomial_logit_lpmf(slice_n_redcards |
                                  n_games[start:end],
                                beta[1] + beta[2] * rating[start:end]);
  }
}
data {
  int<lower=0> N;
  int<lower=0> n_redcards[N];
  int<lower=0> n_games[N];
  vector[N] rating;
  int<lower=1> grainsize;
}
parameters {
  vector[2] beta;
}
model {

  beta[1] ~ normal(0, 10);
  beta[2] ~ normal(0, 1);

  target += reduce_sum(partial_sum_lp, n_redcards, grainsize,
                       n_games, rating, beta);
}
```

#### Current Output:
```
...
stan/lib/stan_math/stan/math/prim/functor/reduce_sum.hpp:83:20: error: void value not ignored as it ought to be
       sum_ += apply(
               ~~~~~^
           [&](auto&&... args) {
           ~~~~~~~~~~~~~~~~~~~~~
             return ReduceFunction()(sub_slice, r.begin(), r.end() - 1, &msgs_,
             ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
                                     args...);
                                     ~~~~~~~~~
           },
           ~~        
           args_tuple_);
           ~~~~~~~~~~~~
mingw32-make.exe: *** [make/program:61: C:\Windows\TEMP\Rtmpwz25Zn\model-30c43d97d48.exe] Error 1
Error: An error occured during compilation! See the message above for more information.
```

#### Expected Output:
`the _lp suffix is reserved and cannot be used with reduce_sum. Please change your function name.` (or something)

#### Current Version:
v2.26.1

**Comment by rok-cesnovar** [comment]
With a recent refactor of handling suffixes by @nhuurre this now errors in stanc3 and would not error in the C++ compile phase, but the error is a bit misleading or weird.

```
Ill-typed arguments supplied to function 'reduce_sum'. Available arguments:
(T[], int, int, ...) => real, T[], int, ...
(T[,], int, int, ...) => real, T[,], int, ...
(T[,,], int, int, ...) => real, T[,,], int, ...
(T[,,,], int, int, ...) => real, T[,,,], int, ...
(T[,,,,], int, int, ...) => real, T[,,,,], int, ...
(T[,,,,,], int, int, ...) => real, T[,,,,,], int, ...
(T[,,,,,,], int, int, ...) => real, T[,,,,,,], int, ...
Where T is any one of int, real, vector, row_vector or matrix.
Instead supplied arguments of incompatible type: (array[] int, int, int, array[] int, vector, vector) => real, array[] int, int, array[] int, vector, vector
```

**Comment by nhuurre** [comment]
I have a branch ready that makes this error (and many others) significantly more intelligible but I didn't want to open a feature pull five minutes before the feature freeze.

This is still a documentation issue. _lp functions are barely mentioned in the docs.

**Comment by rok-cesnovar** [comment]
I wouldnt mind such a PR, its probably considered a bugfix anyways (at least its definitely more of a bugfix than a new feature) . But if you prefer we can also do it after the release. This particular issue is probably very uncommon anyways and at least it errors earlier now.

> This is still a documentation issue. _lp functions are barely mentioned in the docs.

Agreed, though it becomes much less of a pressing docs issue if more details are given in the error message. But lp functions are not documented thoroughly that is for sure.

**Comment by nhuurre** [comment]
It's a pretty big refactor, and the release candidate had some real performance bugs to sort out, but sure if someone wants to review it let's call it a "bugfix". I've been putting it off for too long anyway.

**Comment by WardBrian** [comment]
Resolved by #907
